cleverhans/attacks_tf.py

```diff
@@ -133,275 +133,248 @@ def vatm(model, x, logits, eps, num_iterations=1, xi=1e-6,
-    :param grads_other: a matrix where every element is the sum of forward
-                        derivatives over all non-target classes at that index
-    :param search_domain: the set of input indices that we are considering
-    :param increase: boolean; true if we are increasing pixels, false otherwise
-    :return: (i, j, search_domain) the two input indices selected and the
-             updated search domain
-    """
-    # Compute the size of the input (the number of features)
-    nf = len(grads_target)
-
-    # Remove the already-used input features from the search space
-    invalid = list(set(range(nf)) - search_domain)
-    increase_coef = (2 * int(increase) - 1)
-    grads_target[invalid] = - increase_coef * np.max(np.abs(grads_target))
-    grads_other[invalid] = increase_coef * np.max(np.abs(grads_other))
-
-    # Create a 2D numpy array of the sum of grads_target and grads_other
-    target_sum = grads_target.reshape((1, nf)) + grads_target.reshape((nf, 1))
-    other_sum = grads_other.reshape((1, nf)) + grads_other.reshape((nf, 1))
-
-    # Create a mask to only keep features that match saliency map conditions
-    if increase:
-        scores_mask = ((target_sum > 0) & (other_sum < 0))
+    xshape = tf.shape(x)
+    n = xshape[0]
+    target = tf.cond(tf.equal(0, tf.rank(y)),
+                     lambda: tf.zeros([n], dtype=tf.int32)+y,
+                     lambda: y)
+
+    if isinstance(epochs, float):
+        tmp = tf.to_float(tf.size(x[0])) * epochs
+        epochs = tf.to_int32(tf.floor(tmp))
+
+    if pair:
+        _jsma_fn = _jsma2_impl
     else:
-        scores_mask = ((target_sum < 0) & (other_sum > 0))
+        _jsma_fn = _jsma_impl
 
-    # Create a 2D numpy array of the scores for each pair of candidate features
-    scores = scores_mask * (-target_sum * other_sum)
+    def _fn(i):
+        # `xi` is of the shape (1, ....), the first dimension is the number of
+        # samples, 1 in this case.  `yi` is just a scalar, denoting the target
+        # class index.
+        xi = tf.gather(x, [i])
+        yi = tf.gather(target, i)
 
-    # A pixel can only be selected (and changed) once
-    np.fill_diagonal(scores, 0)
+        # `xadv` is of the shape (1, ...), same as xi.
+        xadv = _jsma_fn(model, xi, yi, epochs=epochs, eps=eps,
+                        clip_min=clip_min, clip_max=clip_max,
+                        min_proba=min_proba)
+        return xadv[0]
 
-    # Extract the best two pixels
-    best = np.argmax(scores)
-    p1, p2 = best % nf, best // nf
+    return tf.map_fn(_fn, tf.range(n), dtype=tf.float32,
+                     back_prop=False, name='jsma_batch')
 
-    # Remove used pixels from our search domain
-    search_domain.discard(p1)
-    search_domain.discard(p2)
 
-    return p1, p2, search_domain
+def _jsma_impl(model, xi, yi, epochs, eps=1.0, clip_min=0.0, clip_max=1.0,
+               min_proba=1.1):
 
+    def _cond(x_adv, epoch, pixel_mask):
```

Comment:
Can you document this function, at least briefly? My understanding is that it checks for two things: (1) the target label is achieved, (2) with enough confidence (i.e., min_proba < proba)